### ISSUE TITLE
How can I know the criterion(Box or Mask) and corresponding epoch for the best.pt in YOLOv8-seg?

### ISSUE THREAD

**Issue by j-ys** [ultralytics/ultralytics#7630]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I successfully trained my custom dataset using the YOLOv8-seg model, and I have obtained the best.pt, last.pt, and result.csv files.
I would like to understand which epoch is the best.pt corresponds to and what criteria are used to determine that result(e.g. fitness about Box or Mask). While I've looked into related issues, I couldn't find information on the specific standards used for saving the best.pt file; it's mentioned that it's saved  **on every maximum fitness epoch**.

So, I found 'How to calculate fitness' in [here (line 741).](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/metrics.py)

I wrote my own code to obtain the epoch with the maximum fitness using the result.csv file.
```javascript
def fitness(x): 
     # Model fitness as a weighted combination of metrics 
     w = [0.1, 0.9]  # weights for [mAP@0.5(B or M), mAP@0.5:0.95(B or M)] 
     return (x[:, :2] * w).sum(1)

# B: Box, M:Mask
selected_column_B = ['metrics/mAP50(B)','metrics/mAP5095(B)']
selected_column_M = ['metrics/mAP50(M)','metrics/mAP5095(M)']

# add columns
df['fitness_B'] = fitness(df[selected_column_B].values)
df['fitness_M'] = fitness(df[selected_column_M].values)

max_index_B = df['fitness_B'].idxmax() #max fitness value
max_index_M = df['fitness_M'].idxmax()

fitness_epoch_B = df.loc[max_index_B, 'epoch']
fitness_epoch_M = df.loc[max_index_M, 'epoch']

print(f"fitness_epoch_B: {fitness_epoch_B}, max_fitness_B: {df['fitness_B'].max()}")
print(f"fitness_epoch_M: {fitness_epoch_M}, max_fitness_M: {df['fitness_M'].max()}")
```
Result:

fitness_epoch_B: 287, max_fitness_B: 0.9637340000000001
fitness_epoch_M: 246, max_fitness_M: 0.929741

So, 
**1) Why is there a difference in the 'max epoch' between 'max_fitness_B' and 'max_fitness_M'?
2) Is the criterion for creating the best.pt file 'max_fitness_M' or 'max_fitness_B'?(Yolov8-seg)**

**Comment by DrawingProcess** [comment]
I have the same question. 

For segmentation task, which one do you **calculate best.pt** based on the mAP of **bbox or mask?**
And the only way to pick the **best.pt epoch** at that time as mentioned above?

**Comment by glenn-jocher** [comment]
@DrawingProcess for YOLOv8-seg, the `best.pt` model is determined by the highest fitness score, which is a weighted combination of both bounding box and mask mAP metrics. The specific weights and criteria used to calculate the fitness score can be found in our documentation. The epoch corresponding to `best.pt` is the one where this maximum fitness score was achieved during training. If you need to identify the exact epoch, you can refer to the `results.csv` file, which logs the fitness scores for each epoch. The method you've used to calculate the fitness from `results.csv` is correct, and you can use it to find the epoch with the highest fitness for both bounding boxes and masks. However, the `best.pt` file is saved based on the overall fitness, which includes both criteria. For more details on the fitness calculation and model selection, please refer to our documentation.

**Comment by DrawingProcess** [comment]
As you mentioned above @j-ys , the highest fitness score of the bbox and mask is different. 
What formula use to calculate best.pt ?

**Comment by glenn-jocher** [comment]
@DrawingProcess The `best.pt` is determined by the highest combined fitness score, which takes into account both bounding box (bbox) and mask metrics. The exact formula for the combined fitness score is a weighted sum of the mAP@0.5 and mAP@0.5:0.95 for both bbox and mask. The weights and the combination method are detailed in our documentation. The epoch that yields the highest combined fitness score across both criteria is the one that `best.pt` corresponds to.

**Comment by DrawingProcess** [comment]
@glenn-jocher Thank you for answer.

I found a [hyperparameter_tuning.md](./docs/en/guides/hyperparameter_tuning.md)
```python
from ultralytics import YOLO

 # Initialize the YOLO model
model = YOLO('yolov8n.pt')

# Tune hyperparameters on COCO8 for 30 epochs
model.tune(data='coco8.yaml', epochs=30, iterations=300, optimizer='AdamW', plots=False, save=False, val=False)
```

If you use model.tune, you can expect to get the yaml file below, so I'll try this. 
```yaml
      # 558/900 iterations complete ✅ (45536.81s)
      # Results saved to /usr/src/ultralytics/runs/detect/tune
      # Best fitness=0.64297 observed at iteration 498
      # Best fitness metrics are {'metrics/precision(B)': 0.87247, 'metrics/recall(B)': 0.71387, 'metrics/mAP50(B)': 0.79106, 'metrics/mAP50-95(B)': 0.62651, 'val/box_loss': 2.79884, 'val/cls_loss': 2.72386, 'val/dfl_loss': 0.68503, 'fitness': 0.64297}
      # Best fitness model is /usr/src/ultralytics/runs/detect/train498
      # Best fitness hyperparameters are printed below.

      lr0: 0.00269
      lrf: 0.00288
      momentum: 0.73375
      weight_decay: 0.00015
      warmup_epochs: 1.22935
      warmup_momentum: 0.1525
      box: 18.27875
      cls: 1.32899
      dfl: 0.56016
      hsv_h: 0.01148
      hsv_s: 0.53554
      hsv_v: 0.13636
      degrees: 0.0
      translate: 0.12431
      scale: 0.07643
      shear: 0.0
      perspective: 0.0
      flipud: 0.0
      fliplr: 0.08631
      mosaic: 0.42551
      mixup: 0.0
      copy_paste: 0.0
    ```

**Comment by glenn-jocher** [comment]
@DrawingProcess Yes, using `model.tune` will allow you to perform hyperparameter tuning on your dataset. The process will automatically search for the best hyperparameters over the specified number of epochs and iterations. At the end of the tuning process, you will receive a YAML file with the best-performing hyperparameters, as shown in your example. This file will detail the hyperparameters that resulted in the highest fitness score during the tuning process. You can then use these optimized hyperparameters for training your model to potentially achieve better performance. Good luck with your tuning! 🚀
